pkg/apis/auditregistration/types.go

```diff
@@ -194,3 +219,107 @@ type ServiceReference struct {
 	// +optional
 	Path *string
 }
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// AuditClass is a set of rules that categorize requests
+//
+// This should be considered a highly privileged object, as modifying it
+// will change what is logged.
+type AuditClass struct {
+	metav1.TypeMeta
+	// +optional
+	metav1.ObjectMeta
+
+	// Spec is the spec for the audit class
+	Spec AuditClassSpec
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// AuditClassList is a list of AuditClass items
+type AuditClassList struct {
+	metav1.TypeMeta
+	// +optional
+	metav1.ListMeta
+
+	// List of audit classes
+	Items []AuditClass
+}
+
+// AuditClassSpec is the spec for the audit class
+type AuditClassSpec struct {
+	// RequestSelectors defines a list of RequestSelectors
+	RequestSelectors []RequestSelector
+}
+
+// RequestSelector selects requests by matching on the given attributes. Selectors are
+// used to compose audit classes.
+type RequestSelector struct {
```

Comment:
@pbarker, @tallclair, @liggitt You know labels add orthogonal, domain-specific semantics to k8s resources and it comes in very handy to be able to filter resource lists based on that. I'm not the first to come up with with "hey, how about label selectors in ..." obviously. For example a couple of folks raised [this wrt RBAC](https://github.com/kubernetes/kubernetes/issues/44703 ) and there were some valid concerns and obstacles shared wrt implementing it in this particular domain.   
I'm wondering, whether the same constraints/concerns apply here too? What do you think about amending the struct with label selectors?

Reply:
Labels are tricky because there aren't any restrictions on them, and we want to make sure a malicious user can't go unnoticed by doctoring labels. I could see an argument for having a selector over audit annotations though (the name is unfortunate though, maybe we need labels too?), in which case you could have an admission controller that translated a whitelisted label into an audit annotation. This does create a bit of an ordering problem though, as the level is currently needed before admission runs.

This is out of scope for this PR though.